druid/src/window.rs

```diff
@@ -324,22 +324,33 @@ impl<T: Data> Window<T> {
         match focus {
             FocusChange::Resign => None,
             FocusChange::Focus(id) => Some(id),
-            FocusChange::Next => self
-                .focus
-                .and_then(|id| self.focus_chain().iter().position(|i| i == &id))
-                .map(|idx| {
-                    let next_idx = (idx + 1) % self.focus_chain().len();
-                    self.focus_chain()[next_idx]
-                }),
-            FocusChange::Previous => self
-                .focus
-                .and_then(|id| self.focus_chain().iter().position(|i| i == &id))
+            FocusChange::Next => self.widget_from_focus_chain(true),
+            FocusChange::Previous => self.widget_from_focus_chain(false),
+        }
+    }
+
+    fn widget_from_focus_chain(&self, forward: bool) -> Option<WidgetId> {
+        self.focus.and_then(|focus| {
+            self.focus_chain()
+                .iter()
+                // Find where the focused widget is in the focus chain
+                .position(|id| id == &focus)
                 .map(|idx| {
+                    // Return the id that's next to it in the focus chain
                     let len = self.focus_chain().len();
-                    let prev_idx = (idx + len - 1) % len;
-                    self.focus_chain()[prev_idx]
-                }),
-        }
+                    let new_idx = if forward {
+                        (idx + 1) % len
+                    } else {
+                        (idx + len - 1) % len
+                    };
+                    self.focus_chain()[new_idx]
+                })
+                .or_else(|| {
+                    // If the currently focused widget isn't in the focus chain,
+                    // then we'll just return the first entry of the chain, if any.
```

Comment:
this highlights a definite little rough spot. What should `tab` do if nothing is focused? A reasonable argument is that it should send focus to the first item in the focus chain.

This is related to what should happen when something that isn't in the focus chain is focused, and we end up in this code path.

Realistically, this shouldn't happen. To get here, a widget has to explicitly request next or previous focus. A widget that is doing that has also had the opportunity to register to participate in focus. Doing one without the other feels like maybe an API violation?

This again touches, to me, on the idea of 'ambient focus'. How I interpret this is basically that if no other widget has focus, the root widget does. A `<tab>` from the root widget should focus the next focusable widget, but the root widget itself should not be in the focus chain; it should only receive focus if another widget resigns focus without trying to pass it on, as might happen when someone presses `<esc>`.

So with all that said, I think that this implementation is reasonable. A widget has asked for next/previous focus, so we should focus something.

This would also then work for the root widget, if it had focus; if it sent `focus_next` then that would just grab the first item in the chain.

Given this, should 'previous' maybe take the last item?

Reply:
Regarding `focus_prev` yes I think it should take the last item, I'll change that. The reason is that I just checked that Chrome does work that way, so it's probably what users expect.

As for a widget being focused without registering for focus, I think this is not an API violation and something that is useful to have. It exists in HTML and can be seen in [this example](https://jsfiddle.net/vtjd961k/) I set up. The use case would be a form that has rarely used fields. Then you can have a textbox which can be focused when clicked to change its contents, but in 95% of use cases the user doesn't need to modify it so it doesn't need to be in the tab cycle. However this would make life more difficult for keyboard only users. 

This starts touching upon accessibility in general and being able to tab through buttons and dropdowns etc. I think there's major value in supporting keyboard navigation automatically in all of druid's included widgets. This is a rather large topic though. I think we should start a new issue to talk about this. It's not super clear what the answer is to add any ambient focus support in this PR right now.

Reply:
your logic here makes sense to me; and I think ambient focus makes sense as a separate patch.

I agree about better support for keyboard navigation, and also yes it's a big one.